agent/internal/agent_test.go

```diff
@@ -0,0 +1,116 @@
+package internal
+
+import (
+	"testing"
+
+	"github.com/docker/docker/api/types/container"
+)
+
+func TestNoAddProxy(t *testing.T) {
+	inputEnv := container.Config{}
+	testAgent := agent{}
+
+	inputEnv.Env = []string{
+		"FIRST_VAR=1",
+	}
+
+	ans := append([]string{}, inputEnv.Env...)
+
+	testAgent.addProxy(&inputEnv)
+
+	// InputEnv should not change because we didn't set any environment variables.
+	if v, ok := compareLists(inputEnv.Env, ans); !ok {
+		if len(v.extraA) != 0 {
+			t.Errorf("Extra variables found in Environment: %v", v.extraA)
+		} else {
+			t.Errorf("Missing variables in Environment: %v", v.extraB)
+		}
+	}
+}
+
+func TestAddProxy(t *testing.T) {
+	inputEnv := container.Config{}
+	testAgent := agent{}
+
+	inputEnv.Env = []string{
+		"FIRST_VAR=1",
+	}
+
+	testAgent.Options.HTTPProxy = "192.168.1.1"
+	testAgent.Options.HTTPSProxy = "192.168.1.2"
+	testAgent.Options.FTPProxy = "192.168.1.3"
+	testAgent.Options.NoProxy = "*.test.com"
+
+	ans := append(inputEnv.Env, []string{
+		"HTTP_PROXY=192.168.1.1",
+		"HTTPS_PROXY=192.168.1.2",
+		"FTP_PROXY=192.168.1.3",
+		"NO_PROXY=*.test.com",
+	}...)
+
+	testAgent.addProxy(&inputEnv)
+
+	if v, ok := compareLists(inputEnv.Env, ans); !ok {
+		if len(v.extraA) != 0 {
+			t.Errorf("Extra variables found in Environment: %v", v.extraA)
+		} else {
+			t.Errorf("Missing variables in Environment: %v", v.extraB)
+		}
+	}
+}
+
+func TestAlreadyAddedProxy(t *testing.T) {
+	inputEnv := container.Config{}
+	testAgent := agent{}
+
+	inputEnv.Env = []string{
+		"FIRST_VAR=1",
+		"HTTP_PROXY=10.0.0.1",
+	}
+
+	testAgent.Options.HTTPProxy = "10.0.0.2"
+
+	// We should be overriding earlier proxy variables
```

Comment:
Does this mean that the experiment config variables would be overridden by the agent config variables?

If so, that seems like it would be pretty surprising behavior from the user's perspective.  As the user, I would expect the more specific config (experiment level) to take precedence over the more generic config (agent level) in the same way that the master.yaml can specify a default checkpoint config but each experiment has the ability to override that with its own configuration.

Reply:
Note: changed the behavior to not override experiment config